server/src/main/java/org/opensearch/tasks/TaskResourceTrackingService.java

```diff
@@ -0,0 +1,234 @@
+     * Called when a thread finishes working on a task's runnable.
+     *
+     * @param taskId   of the task for which runnable is complete
+     * @param threadId of the thread which executed the runnable and we need to check resource usage for this thread
+     */
+    @Override
+    public void taskExecutionFinishedOnThread(long taskId, long threadId) {
+        if (resourceAwareTasks.containsKey(taskId)) {
+            resourceAwareTasks.get(taskId).stopThreadResourceTracking(threadId, WORKER_STATS, getResourceUsageMetricsForThread(threadId));
+        }
+    }
+
+    public Map<Long, Task> getResourceAwareTasks() {
+        return Collections.unmodifiableMap(resourceAwareTasks);
+    }
+
+    private ResourceUsageMetric[] getResourceUsageMetricsForThread(long threadId) {
+        ResourceUsageMetric currentMemoryUsage = new ResourceUsageMetric(
+            ResourceStats.MEMORY,
+            threadMXBean.getThreadAllocatedBytes(threadId)
+        );
+        ResourceUsageMetric currentCPUUsage = new ResourceUsageMetric(ResourceStats.CPU, threadMXBean.getThreadCpuTime(threadId));
+        return new ResourceUsageMetric[] { currentMemoryUsage, currentCPUUsage };
+    }
+
+    private boolean isThreadWorkingOnTask(Task task, long threadId) {
+        List<ThreadResourceInfo> threadResourceInfos = task.getResourceStats().getOrDefault(threadId, Collections.emptyList());
+
+        for (ThreadResourceInfo threadResourceInfo : threadResourceInfos) {
+            if (threadResourceInfo.isActive()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean validateNoActiveThread(Task task) {
+        for (List<ThreadResourceInfo> threadResourceInfos : task.getResourceStats().values()) {
+            for (ThreadResourceInfo threadResourceInfo : threadResourceInfos) {
+                if (threadResourceInfo.isActive()) return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Adds Task Id in the ThreadContext.
+     * <p>
+     * Stashes the existing ThreadContext and preserves all the existing ThreadContext's data in the new ThreadContext
+     * as well.
+     *
+     * @param task for which Task Id needs to be added in ThreadContext.
+     * @return StoredContext reference to restore the ThreadContext from which we created a new one.
+     * Caller can call context.restore() to get the existing ThreadContext back.
+     */
+    private ThreadContext.StoredContext addTaskIdToThreadContext(Task task) {
+        ThreadContext threadContext = threadPool.getThreadContext();
+
+        boolean staleIdPresentInThreadContext = threadContext.getTransient(TASK_ID) != null
+            && !resourceAwareTasks.containsKey((long) threadContext.getTransient(TASK_ID));
+
+        if (staleIdPresentInThreadContext) {
+            logger.debug(
+                "Stale Task Id should ideally be not present in thread context. Current task Id: {}, Stale task Id: {}, Thread id: {}",
```

Comment:
This statement isn't really actionable. What should a user do if they see this?

Reply:
Though no direct actionable if someone sees this now. Just for future debugging purposes.

Reply:
The "should ideally" phrasing seems too vague to me. What are the scenarios under which this could happen? If this would only happen due to a bug, then making it info or warn level with a message like "Previous task ID was not removed from thread context..." would be better, or maybe even an `assert`.

Reply:
This could happen if due to some reason AutoCloseable returned by `TaskResourceService` was not closed. Have added a separate test for that. Bumped up the log level to warn